prow/plugins/welcome/welcome.go

```diff
@@ -0,0 +1,91 @@
+/*
+Copyright 2018 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package welcome
+
+import (
+	"fmt"
+
+	"github.com/sirupsen/logrus"
+
+	"k8s.io/test-infra/prow/github"
+	"k8s.io/test-infra/prow/pluginhelp"
+	"k8s.io/test-infra/prow/plugins"
+)
+
+// TODO(bentheelder): unit tests, config, write an actual greeting message
+
+const (
+	pluginName     = "welcome"
+	welcomeMessage = "Welcome!" // TODO, see above
+)
+
+func init() {
+	plugins.RegisterPullRequestHandler(pluginName, handlePullRequest, helpProvider)
+}
+
+func helpProvider(config *plugins.Configuration, enabledRepos []string) (*pluginhelp.PluginHelp, error) {
+	// The {WhoCanUse, Usage, Examples} fields are omitted because this plugin is not triggered with commands.
+	return &pluginhelp.PluginHelp{
+			Description: "The welcome plugin posts a welcoming message when it detects a user's first contribution to a repo.",
+		},
+		nil
+}
+
+type githubClient interface {
+	CreateComment(owner, repo string, number int, comment string) error
+	FindIssues(query, sort string, asc bool) ([]github.Issue, error)
+}
+
+type client struct {
+	GitHubClient githubClient
+	Logger       *logrus.Entry
+}
+
+func getClient(pc plugins.PluginClient) client {
+	return client{
+		GitHubClient: pc.GitHubClient,
+		Logger:       pc.Logger,
+	}
+}
+
+func handlePullRequest(pc plugins.PluginClient, pre github.PullRequestEvent) error {
+	return handlePR(getClient(pc), pre)
+}
+
+func handlePR(c client, pre github.PullRequestEvent) error {
+	// Only consider newly opened PRs
+	if pre.Action != github.PullRequestActionOpened {
+		return nil
+	}
+
+	// search for PRs from the author in this repo
+	org := pre.PullRequest.Base.Repo.Owner.Login
+	repo := pre.PullRequest.Base.Repo.Name
+	user := pre.PullRequest.User.Login
+	query := fmt.Sprintf("is:pr repo:%s/%s author:%s", org, repo, user)
+	issues, err := c.GitHubClient.FindIssues(query, "", false)
```

Comment:
TODO: really we don't even want the issue objects, we just want the totals. We unmarshall this but FindIssues only returns the []Issue, not the total 😕

Reply:
https://developer.github.com/v3/#pagination

Almost seems like we ought to refactor the client to make paginated results return a channel

Reply:
Yeah, another option would be to use the graphQL APIs more instead, but our integration with that seems a bit unwieldy and looks probably more annoying to unit test.

I could also patch this method to return:
https://github.com/kubernetes/test-infra/blob/5ed9c588f9750706a8c84066e8c9e9c95975319f/prow/github/types.go#L428-L432
instead of just the inner `Issues` field...